Move RenderElementType to RenderObject to simplify and devirtualize type checks

rniwa opened the pull request:
#### 5b7421b95b8ca058e421a83a8436f0db859648e3
<pre>
Move RenderElementType to RenderObject to simplify and devirtualize type checks
<a href="https://bugs.webkit.org/show_bug.cgi?id=266531">https://bugs.webkit.org/show_bug.cgi?id=266531</a>

Reviewed by Chris Dumez.

This PR moves OptionSet&lt;RenderElementType&gt; from RenderElement to RenderObject to eliminate
isRenderElement checks from isRenderReplaced, isRenderBoxModelObject, etc...

It also devirtualizes isRenderSVGContainer() and isLegacyRenderSVGContainer() with two more
bit flags: LegacyRenderSVGContainer and RenderSVGContainer. This eliminates the last virtual
function calls in RenderTreeBuilder::detach.

* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::RenderDeprecatedFlexibleBox::RenderDeprecatedFlexibleBox):
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::RenderElement):
(WebCore::RenderElement::createFor):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::typeFlags const): Deleted.
(WebCore::RenderElement::isRenderLayerModelObject const): Deleted.
(WebCore::RenderElement::isRenderBoxModelObject const): Deleted.
(WebCore::RenderElement::isRenderBlock const): Deleted.
(WebCore::RenderElement::isRenderBlockFlow const): Deleted.
(WebCore::RenderElement::isRenderReplaced const): Deleted.
(WebCore::RenderElement::isRenderInline const): Deleted.
(WebCore::RenderElement::isRenderFlexibleBox const): Deleted.
(WebCore::RenderElement::isRenderTextControl const): Deleted.
(WebCore::RenderObject::isRenderLayerModelObject const): Deleted.
(WebCore::RenderObject::isRenderBoxModelObject const): Deleted.
(WebCore::RenderObject::isRenderBlock const): Deleted.
(WebCore::RenderObject::isRenderBlockFlow const): Deleted.
(WebCore::RenderObject::isRenderReplaced const): Deleted.
(WebCore::RenderObject::isRenderInline const): Deleted.
(WebCore::RenderObject::isRenderFlexibleBox const): Deleted.
(WebCore::RenderObject::isRenderTextControl const): Deleted.
(WebCore::RenderObject::isFlexibleBoxIncludingDeprecated const): Deleted.
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::RenderObject):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderReplaced const):
(WebCore::RenderObject::isRenderBoxModelObject const):
(WebCore::RenderObject::isRenderBlock const):
(WebCore::RenderObject::isRenderBlockFlow const):
(WebCore::RenderObject::isRenderInline const):
(WebCore::RenderObject::isRenderLayerModelObject const):
(WebCore::RenderObject::isRenderTextControl const):
(WebCore::RenderObject::isRenderTextControlSingleLine const):
(WebCore::RenderObject::isRenderSVGContainer const):
(WebCore::RenderObject::isLegacyRenderSVGContainer const):
(WebCore::RenderObject::isRenderDeprecatedFlexibleBox const):
(WebCore::RenderObject::isRenderFlexibleBox const):
(WebCore::RenderObject::isFlexibleBoxIncludingDeprecated const):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::RenderText):
* Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
(WebCore::RenderTextControlSingleLine::RenderTextControlSingleLine):
* Source/WebCore/rendering/RenderTextControlSingleLine.h:
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::RenderSVGContainer):
* Source/WebCore/rendering/svg/RenderSVGContainer.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::RenderSVGModelObject):
* Source/WebCore/rendering/svg/RenderSVGModelObject.h:
(WebCore::RenderSVGModelObject::RenderSVGModelObject):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::LegacyRenderSVGContainer):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::LegacyRenderSVGModelObject):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.h:
(WebCore::LegacyRenderSVGModelObject::LegacyRenderSVGModelObject):

Canonical link: <a href="https://commits.webkit.org/272195@main">https://commits.webkit.org/272195@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/793a47884a195c113532fb4c25e40111cc3a3b2b

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/30867 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/9545 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/32540 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/33369 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/27883 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/31619 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/11866 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/6795 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/27754 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/31186 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/8020 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/27608 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/6882 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/7036 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/27486 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/34707 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/28093 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/27967 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/33189 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/7082 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/5158 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/31022 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/8793 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/7794 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/4003 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/7636 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

annevk commented on Source/WebCore/rendering/RenderObject.h:
```diff
@@ -1206,7 +1223,8 @@ class RenderObject : public CachedImageClient, public CanMakeCheckedPtr {
 
     SingleThreadWeakPtr<RenderElement> m_parent;
-    SingleThreadWeakPtr<RenderObject> m_previous;
+    SingleThreadPackedWeakPtr<RenderObject> m_previous;
```
Did you mean to make this change in this PR? It isn't mentioned in the commit message.

rniwa replied:
Yes, this is intentional as we'd increase the size of RenderObject otherwise.

cdumez reviewed (approved):
LGTM

rniwa committed:
Move RenderElementType to RenderObject to simplify and devirtualize type checks
https://bugs.webkit.org/show_bug.cgi?id=266531

Reviewed by Chris Dumez.

This PR moves OptionSet<RenderElementType> from RenderElement to RenderObject to eliminate
isRenderElement checks from isRenderReplaced, isRenderBoxModelObject, etc...

It also devirtualizes isRenderSVGContainer() and isLegacyRenderSVGContainer() with two more
bit flags: LegacyRenderSVGContainer and RenderSVGContainer. This eliminates the last virtual
function calls in RenderTreeBuilder::detach.

* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.cpp:
(WebCore::RenderDeprecatedFlexibleBox::RenderDeprecatedFlexibleBox):
* Source/WebCore/rendering/RenderDeprecatedFlexibleBox.h:
* Source/WebCore/rendering/RenderElement.cpp:
(WebCore::RenderElement::RenderElement):
(WebCore::RenderElement::createFor):
* Source/WebCore/rendering/RenderElement.h:
(WebCore::RenderElement::typeFlags const): Deleted.
(WebCore::RenderElement::isRenderLayerModelObject const): Deleted.
(WebCore::RenderElement::isRenderBoxModelObject const): Deleted.
(WebCore::RenderElement::isRenderBlock const): Deleted.
(WebCore::RenderElement::isRenderBlockFlow const): Deleted.
(WebCore::RenderElement::isRenderReplaced const): Deleted.
(WebCore::RenderElement::isRenderInline const): Deleted.
(WebCore::RenderElement::isRenderFlexibleBox const): Deleted.
(WebCore::RenderElement::isRenderTextControl const): Deleted.
(WebCore::RenderObject::isRenderLayerModelObject const): Deleted.
(WebCore::RenderObject::isRenderBoxModelObject const): Deleted.
(WebCore::RenderObject::isRenderBlock const): Deleted.
(WebCore::RenderObject::isRenderBlockFlow const): Deleted.
(WebCore::RenderObject::isRenderReplaced const): Deleted.
(WebCore::RenderObject::isRenderInline const): Deleted.
(WebCore::RenderObject::isRenderFlexibleBox const): Deleted.
(WebCore::RenderObject::isRenderTextControl const): Deleted.
(WebCore::RenderObject::isFlexibleBoxIncludingDeprecated const): Deleted.
* Source/WebCore/rendering/RenderObject.cpp:
(WebCore::RenderObject::RenderObject):
* Source/WebCore/rendering/RenderObject.h:
(WebCore::RenderObject::isRenderReplaced const):
(WebCore::RenderObject::isRenderBoxModelObject const):
(WebCore::RenderObject::isRenderBlock const):
(WebCore::RenderObject::isRenderBlockFlow const):
(WebCore::RenderObject::isRenderInline const):
(WebCore::RenderObject::isRenderLayerModelObject const):
(WebCore::RenderObject::isRenderTextControl const):
(WebCore::RenderObject::isRenderTextControlSingleLine const):
(WebCore::RenderObject::isRenderSVGContainer const):
(WebCore::RenderObject::isLegacyRenderSVGContainer const):
(WebCore::RenderObject::isRenderDeprecatedFlexibleBox const):
(WebCore::RenderObject::isRenderFlexibleBox const):
(WebCore::RenderObject::isFlexibleBoxIncludingDeprecated const):
* Source/WebCore/rendering/RenderText.cpp:
(WebCore::RenderText::RenderText):
* Source/WebCore/rendering/RenderTextControlSingleLine.cpp:
(WebCore::RenderTextControlSingleLine::RenderTextControlSingleLine):
* Source/WebCore/rendering/RenderTextControlSingleLine.h:
* Source/WebCore/rendering/svg/RenderSVGContainer.cpp:
(WebCore::RenderSVGContainer::RenderSVGContainer):
* Source/WebCore/rendering/svg/RenderSVGContainer.h:
* Source/WebCore/rendering/svg/RenderSVGModelObject.cpp:
(WebCore::RenderSVGModelObject::RenderSVGModelObject):
* Source/WebCore/rendering/svg/RenderSVGModelObject.h:
(WebCore::RenderSVGModelObject::RenderSVGModelObject):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.cpp:
(WebCore::LegacyRenderSVGContainer::LegacyRenderSVGContainer):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGContainer.h:
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.cpp:
(WebCore::LegacyRenderSVGModelObject::LegacyRenderSVGModelObject):
* Source/WebCore/rendering/svg/legacy/LegacyRenderSVGModelObject.h:
(WebCore::LegacyRenderSVGModelObject::LegacyRenderSVGModelObject):

Canonical link: https://commits.webkit.org/272195@main
